### ISSUE TITLE
Making user data requests after receiving tokens from the authorization process

### ISSUE THREAD

**Issue by alehuo** [OpenIDC/pyoidc#732]
I'm currently implementing user authentication using OpenID Connect and pyoidc library. I'm storing the access token and refresh token in the user's session.

As specified in the OpenID Connect specification, I can do an user data request to the `/userdata` endpoint of my authentication provider with the pyoidc provided access token.

How are user data requests made with pyoidc after the user has been successfully authenticated?

I've tried to follow the examples provided in the repository, but the data requests are made in such a manner that makes it pretty hard to get right in the first try.

This is my current piece of code:

```python
session['state'] = rndstr();
dataRequest = client.construct_UserInfoRequest(None, request_args = {
            "token": session['token']
        }, extra_args=None, scope = "openid offline_access email", state=session["state"])
```

After running this example, I'm getting ```oic.oauth2.exception.GrantError: No grant found for state:'XXXXXXX'```

I can contribute to the repo a few easy examples about clients when I get this sorted out first. I'm not very familiar with OpenID Connect standard, so my code example might not be the prettiest.

**Comment by schlenk** [comment]
The code assumes you use the same client for authorization and access of protected resources like userinfo in a few spots and stores access tokens etc. in some internal dictionary. The key to that dictionary is usually the `state` parameter you used during the authorization. So your error here is trying to access this grant cache with a random state.

Usually you would start from `client.do_user_info_request()` to do all the hard lifting for you, instead of constructing the request yourself. If you want to serialize the client state, e.g. in the users session, you might have a look at the `oic.oic.consumer.Consumer` class, which makes the client state easier to serialize and restore.

See the end of this example for a userinfo request, it does not yet use the new consumer code, but should get you started:
https://pyoidc.readthedocs.io/en/latest/examples/rp.html#authorization-code-flow

**Comment by alehuo** [comment]
I started using Consumer for the connections and seems everything is working nicely. Thanks for your help @schlenk !

I experimented on a custom Redis-based session backend and got an error when creating a custom SessionBackend:
```
TypeError: Object of type 'KeyJar' is not JSON serializable
```
Seems that KeyJar cannot be serialized with json or pickle libraries. I will open a new issue to discuss this further, if applicable.
